Java/core/src/main/java/com/amazon/randomcutforest/store/SmallIndexManager.java

```diff
@@ -125,8 +182,26 @@ protected short takeIndex() {
      * @param index The index value to release.
      */
     protected void releaseIndex(int index) {
-        checkValidIndex(index);
+        checkArgument(index >= 0 && index < capacity, " incorrect index to release");
+        checkArgument(occupied.get(index), " releasing the index twice");
         occupied.clear(index);
+        if (freeIndexPointer + 1 >= freeIndexes.length) {
+            if (index == capacity - (freeIndexPointer + 1) - 1) {
+                ++freeIndexPointer;
+                return;
+            } else {
+                // for most of the applications of this small sampler we would need one extra
+                // entry; but freeIndexPointer can be -1 as well
+                // this can be changed to a larger value to reduce repeated
+                // allocation/deallocation
+                int cap = Math.min(capacity, freeIndexPointer + 2);
+                int oldLength = freeIndexes.length;
+                freeIndexes = Arrays.copyOf(freeIndexes, cap);
+                for (int j = oldLength; j < cap && j < freeIndexPointer + 1; j++) {
+                    freeIndexes[j] = (short) (capacity - j - 1);
```

Comment:
Why are we populating these values? Won't they be overwritten by future calls to `releaseIndex`? When someone calls `releaseIndex` for a given index, does that imply that they already released these other indexes so they should be added to the free index stack?

Reply:
We are populating them because the newly added "index" breaks the implicit guarantee -- and if we add the break to the pattern then we need to ensure that the intermediate values are filled in as well. We are adding the index to the stack at the last line of the method.